stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -187,63 +187,102 @@ column. In conjunction with [`nonzeros`](@ref) and
+        s = min(2maxWidth / n, 4maxHeight / m)
+        scaleHeight = floor(Int, s * m)
+        scaleWidth = floor(Int, s * n)
+    else
+        scaleHeight = m
+        scaleWidth = n
     end
 
-    function _get_cols(from, to)
-        idx = eltype(getcolptr(S))[]
-        c = searchsortedlast(getcolptr(S), from)
-        for i = from:to
-            while i == getcolptr(S)[c+1]
-                c +=1
-            end
-            push!(idx, c)
+    # `brailleGrid` is used to store the needed braille characters for
+    # the matrix `S`. Each row of the braille pattern to print is stored
+    # in a column of `brailleGrid`.
+    brailleGrid = fill(UInt16(10240), (scaleWidth - 1) ÷ 2 + 2, (scaleHeight - 1) ÷ 4 + 1)
+    brailleGrid[end, :] .= '\n'
+
+    rvals = rowvals(S)
+    rowscale = max(1, scaleHeight - 1) / max(1, m - 1)
+    colscale = max(1, scaleWidth - 1) / max(1, n - 1)
+    @inbounds for j = 1:n
+        # Scale the column index `j` to the best matching column index
+        # of a matrix of size `scaleHeight × scaleWidth`
+        sj = round(Int, (j - 1) * colscale + 1)
+        for x in nzrange(S, j)
+            # Scale the row index `i` to the best matching row index
+            # of a matrix of size `scaleHeight × scaleWidth`
+            si = round(Int, (rvals[x] - 1) * rowscale + 1)
+
+            # Given the index pair `(si, sj)` of the scaled matrix,
+            # calculate the corresponding triple `(k, l, p)` such that the
+            # element at `(si, sj)` can be found at position `(k, l)` in the
+            # braille grid `brailleGrid` and corresponds to the 1-dot braille
+            # character `brailleBlocks[p]`
+            k = (sj - 1) ÷ 2 + 1
+            l = (si - 1) ÷ 4 + 1
+            p = ((sj - 1) % 2) * 4 + ((si - 1) % 4 + 1)
+
+            brailleGrid[k, l] |= brailleBlocks[p]
         end
-        idx
     end
+    foreach(c -> print(io, Char(c)), @view brailleGrid[1:end-1])
+end
 
-    rows = displaysize(io)[1] - 4 # -4 from [Prompt, header, newline after elements, new prompt]
-    if !get(io, :limit, false) || rows >= nnz(S) # Will the whole matrix fit when printed?
-        cols = _get_cols(1, nnz(S))
-        padr, padc = ndigits.((maximum(rowvals(S)[1:nnz(S)]), cols[end]))
-        _format_line.(1:nnz(S), cols, padr, padc)
-    else
-        if rows <= 2
-            print(io, "\n  \u22ee")
-            return
-        end
-        s1, e1 = 1, div(rows - 1, 2) # -1 accounts for \vdots
-        s2, e2 = nnz(S) - (rows - 1 - e1) + 1, nnz(S)
-        cols1, cols2 = _get_cols(s1, e1), _get_cols(s2, e2)
-        padr = ndigits(max(maximum(rowvals(S)[s1:e1]), maximum(rowvals(S)[s2:e2])))
-        padc = ndigits(cols2[end])
-        _format_line.(s1:e1, cols1, padr, padc)
-        print(io, "\n  \u22ee")
-        _format_line.(s2:e2, cols2, padr, padc)
+function Base.show(io::IOContext, S::AbstractSparseMatrixCSC)
+    if size(S, 1) <= 20 && size(S, 2) <= 20
```

Comment:
Use `displaysize` to determine this cutoff?

Reply:
This threshold (currently set to 20) was introduced because displaying small sparse matrices with braille patterns isn't really helpful. If the matrix is still to big to be printed on the screen, the `print_matrix` function already wraps it. That's in my opinion a better solution than displaying such small matrices with braille patterns. Also note that making this dependent on `displaysize` would imply that two users would see a sparse matrix `sprandn(10, 10, 0.4)` differently depending on their screen size. I'm not sure if this is a good idea.

Maybe some other opinions on this would be helpful?

Reply:
Right, the cutoff isn't determined by the size of the display, it's determined by the minimum size for braille output to be comprehensible.

I picked `20` out of a hat, however, so one might want to experiment a bit to see how things look.   Maybe show the braille output in a comment for a bunch of small matrix sizes and we can judge how they look?

Reply:
Here are some examples of a structured `n x n` matrix in different sizes printed with braille patterns:
![image](https://user-images.githubusercontent.com/1856425/69791492-ec10af00-11c4-11ea-85ae-f3511ebb9bf1.png)

I think `20` is a reasonable choice for that cutoff value.

Reply:
16x16 would probably be fine too.

Reply:
We might change the condition for non-braille printing to
```jl
if max(size(S)...) < 16 && !get(io, :compact, false)
```
so that a `compact` IOContext ensures braille display.